experimental/oauther/oauth2.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package oauther
+
+import (
+	"encoding/json"
+
+	"github.com/gorilla/mux"
+	"github.com/mattermost/mattermost-plugin-api/experimental/bot/logger"
+	"github.com/mattermost/mattermost-server/v5/plugin"
+	"golang.org/x/oauth2"
+)
+
+const (
+	oAuth2StateTimeToLive = 300 // seconds
+)
+
+type OAuther interface {
+	GetToken(userID string) (*oauth2.Token, error)
+	GetURL() string
+	Deauth(userID string) error
+}
+
+type oAuther struct {
+	api             plugin.API
+	onConnect       func(userID string, token *oauth2.Token)
+	storePrefix     string
+	pluginURL       string
+	oAuthURL        string
+	connectedString string
+	config          *oauth2.Config
+	logger          logger.Logger
+}
+
+func NewOAuther(
+	r *mux.Router,
+	api plugin.API,
+	pluginURL,
+	oAuthURL,
+	storePrefix,
+	connectedString string,
+	onConnect func(userID string, token *oauth2.Token),
+	oAuthConfig *oauth2.Config,
+	loggerBot logger.Logger,
+) OAuther {
+	o := &oAuther{
+		api:             api,
+		onConnect:       onConnect,
+		storePrefix:     storePrefix,
+		pluginURL:       pluginURL,
+		oAuthURL:        oAuthURL,
+		config:          oAuthConfig,
+		connectedString: connectedString,
+		logger:          loggerBot,
+	}
+
+	o.config.RedirectURL = pluginURL + oAuthURL + "/complete"
+
+	oauth2Router := r.PathPrefix(oAuthURL).Subrouter()
```

Comment:
Instead of accepting a router and binding the client to `gorilla/mux`, can we make oAuther an `http.Handler` instead?

Reply:
I would like to get rid of gorilla/mux and add something more abstract, so gorilla/mux or any other service could be used for this. Nevertheless, I was not able to think about a better way to do this, and gorilla/mux is broadly used among our plugins.

Reply:
5/5 that I personally dislike gorilla/mux, specifically for its lack of support for request-time "variables" (context) - they resort to the go context which is the wrong thing to do for the use-case.

Moving the dependency to the caller by providing an http.Handler sounds like a good approach.

Reply:
I also did this, but not 100% sure I did what was expected. Any feedback on these changes would be welcomed.